### ISSUE TITLE
Change the WebDriver instance on the BrowseTheWeb Ability

### ISSUE THREAD

**Issue by Maelgwyn** [q2ebanking/boa-constrictor#101]
I have to run tests in a flakey environment. I know I shouldn't but it is what it is.

During login I occasionally get the "The HTTP request to the remote WebDriver server for URL timed out after 60 seconds" error and attempts to re-navigate trying to recover from this fail.

I'd like to be able to close the driver, start a new one and update it on the BrowseTheWeb ability or remove the BrowseTheWeb ability and readd with the new driver.

The Webdriver on BrowseTheWeb has no setter.

Am I missing an easier way of doing this?

**Comment by pl-aknight** [comment]
Hi @Maelgwyn! Thanks for your question.

I'm sorry to hear about flaky tests and WebDriver timeouts. I feel your pain. I know what that's like.

I have some questions:

- What browser types and operating systems do your tests use?
- Are you using local browsers, Selenium Grid, or a cloud-based service like Sauce Labs?
- Does the WebDriver instance successfully complete any interactions, such as loading the page or sending keystrokes?
- Are you sure the problem is WebDriver, or is the app under test just super slow sometimes?

In Boa Constrictor's current design, the `WebDriver` property for the `BrowseTheWeb` Ability is read-only. The reason for this design choice is to prevent callers from modifying the Actor's WebDriver. Changing the WebDriver object during testing could be dangerous. In the same sense, Actors cannot have Abilities removed dynamically. Therefore, I'm hesitant to allow callers to change the Actor's WebDriver instance.

In my current test automation project that uses Boa Constrictor + SpecFlow, I take a different approach: I let tests fail whenever WebDriver crashes. I also configure SpecFlow to automatically rerun any failed tests twice to determine if the failure is intermittent or consistent. A WebDriver instance could theoretically crash at any time during a test. The only time it could be recovered is right at the beginning, and that requires custom retry logic. Relying on SpecFlow's retry logic makes my automation code simpler, and it provides data for intermittencies.

Nevertheless, I see value in the approach you desire to take. Let me think on it overnight. I'll try to come up with an elegant solution. In the meantime, please answer the questions above. Thanks!

**Comment by Maelgwyn** [comment]
Thanks for the swift reply.

**Comment by Maelgwyn** [comment]
- Currently we're limiting to just ChromeDriver on Windows.
- It's a local browser. No Grid, Sauce Labs or Browser Stack yet.
- In failing tests no interactions pass, the failure is on the first attempt to navigate.
- It could well be that the app is just super slow sometimes.

I have liked using SpecFlow in the past but with the Boa test being so readable and the limited audience for the tests it seemed a bit OTT in this instance. I could change my mind though.

Appreciate you taking the time to consider this

**Comment by pl-aknight** [comment]
One time in the past, I had an issue where local ChromeDriver on Windows 10 would sometimes arbitrarily freeze if I did not wait up to 5-10 seconds after initialization before sending commands to it. That was when ChromeDriver's version was in the 70s and 80s. It no longer happens for me. I could never figure out why the problem happened (and also why it stopped).

**Comment by pl-aknight** [comment]
There are a few ways we could handle this:

1. We could stick to the original design, not allowing for the WebDriver instance to be changed after setting it. The only way to try a fresh WebDriver would be to construct a new Actor.
2. We could add a public setter to `BrowseTheWeb`'s `WebDriver` property. This would enable any caller to directly change the WebDriver in the Ability at any time. Unfortunately, there would be no automatic logging for this change. Any problems resulting from this change would be difficult to trace.
3. We could add an internal setter to `BrowseTheWeb`'s `WebDriver` property and create a new Task to change the WebDriver instance. Using a Task would make the call to change the WebDriver instance more readable in the code, and it would also provide automatic logging. If desired, the Task could also attempt to quit the old WebDriver being replaced.
4. We could enable the Actor to remove Abilities. In the present implementation, callers can add new Abilities to Actors, but they cannot remove Abilities from Actors. Abilities become permanent. This is good for test automation but not necessarily good for larger system usage. To replace a WebDriver instance, a caller could remove the old `BrowseTheWeb` Ability and add a new one with the new WebDriver instance. Unfortunately, calls would feel a bit clunky (Actor.Remove --> Actor.Can), and dynamically changing Abilities could dangerously cause confusing behavior.

Out of all these options, I think I favor the third.

**Comment by Maelgwyn** [comment]
I'm all aboard for options three.

**Comment by pl-aknight** [comment]
Members from my team agree. I think we'll implement it.

**Comment by pl-aknight** [comment]
@Maelgwyn, we just released Boa Constrictor 0.14.0. You can now do this:

```csharp
// Does not quit old WebDriver
Actor.AttemptsTo(ChangeWebDriver.To(newWebDriver));

// Attempts to quit old WebDriver first
Actor.AttemptsTo(ChangeWebDriver.To(newWebDriver).AfterQuittingOldWebDriver());
```

**Comment by Maelgwyn** [comment]
Awesome, thanks so much.
